### PR TITLE
refactor mutiprocess updater hyperparam tuning

### PR DESCRIPTION
**Pull request by emcastillo**
#7481 related

avoids hardcoding of optimizers in the `MultiprocessParallelUpdater` to tune
some hyperparameters according to the number of devices.

Provides an interface in the optimizer class to do this update

### CODE REVIEW

File: chainer/optimizer.py [hunk 1]
```diff
@@ -712,6 +712,15 @@ def update_loss_scale(self):
         self._loss_scale = max(1, min(self._loss_scale_max,
                                       self._loss_scale * multiplier))
 
+    def multidevice_hyperparameter_correction(self, num_devices):
```
Comment (niboshi) [inline chainer/optimizer.py]: I suspect the name `multidevice_hyperparameter_correction` is not something we really want to represent.

What we want to do is "to adapt the hyperparameters for modified batch size".
It's better to come up with a name that express that.

`adjust_hyperparameters_for_modified_batch_size`?
Or perhaps simply `adjust_hyperparameters`? But I think explicitness is more important here.

Also, the argument should be a real number value, like `factor` (= `1 / num_devices`).

Comment (niboshi) [reply chainer/optimizer.py]: Or how about introducing something like `Optimizer.batch_size_factor` and `UpdateRule.batch_size_factor` and referring to the latter from `UpdateRule.update_core()`?
I think it must be implemented as a hyperparameter in order for it to be correctly propagated from `Optimizer` to `UpdateRule`.

Comment (emcastillo) [reply chainer/optimizer.py]: You mean, doing the hyperparameter adjust inside `UpdateRule.update_core()`?

Comment (niboshi) [reply chainer/optimizer.py]: Yes. Not rewriting the hyperparameter indefinitely, but just using an adjusted hyperparameter for each single update.
This will be no-compat because the hyperparameter values exposed to users will change (currently adjusted values are exposed, but after this change, they will be non-adjusted values).

Comment (emcastillo) [reply chainer/optimizer.py]: One of the things that I see in this case, is that we need to write explicit UpdateRules for all the optimizers. In the case of Adam, the derived optimizers share the `AdamUpdateRule`, but they adjust the hyper-parameters differently.

I am not against this change as a think the potential benefits compensate the extra verbosity.

Comment (niboshi) [reply chainer/optimizer.py]: By "vebosity" do you mean the necessity to implement separate `UpdateRule`s?
I don't think this is necessary. `AdamRule` can implement different adjustment formulas for different modes (amgrad / adabound etc.) in a single class.
Even if we need separate `UpdateRule` classes, it's not necessarily a bad thing. The code will be more organized.

Comment (emcastillo) [reply chainer/optimizer.py]: Yes, I tried to update the comment and blew it. Today is not my day

Comment (emcastillo) [reply chainer/optimizer.py]: If we go with the hyperparameter approach it seems that we need to pass`batch_size_factor` as a parameter of the optimizer constructor.
This will make the user to explicitly define it when instantiating the optimizer. Do you have any idea about how to solve this?

Comment (niboshi) [reply chainer/optimizer.py]: Hmm, I don't understand what the problem is.
As you said it can be assigned via a property `optimizer.batch_size_factor`, not via constructor.
But what does this part mean?

> then override `update_core` for the `UpdateRule` to set the actual value.

What's "the actual value" referring to? `optimizer.batch_size_factor`?

Comment (emcastillo) [reply chainer/optimizer.py]: the learning rate or eps value adjusted with the batch_size_factor

I feel like this can be done with a regular property to set the eps or the lr instead of a hyperparameter.

Comment (niboshi) [reply chainer/optimizer.py]: Do you mean reassigning calculated hyperparameters as attributes of `UpdateRule`?
I don't think this is needed. We can calculate the adjusted hyperparameters every time `update_core` is called.

Comment (emcastillo) [reply chainer/optimizer.py]: I just added a commit that does it for optimizer.py and ada_grad.py
It is an example and not complete, but I would like to know if this is the right way to do it.
Thanks!

File: chainer/optimizers/ada_grad.py [hunk 2]
```diff
@@ -20,7 +20,7 @@ class AdaGradHyperparameter(tpe.Protocol):
 
 _default_hyperparam = optimizer.Hyperparameter()  # type: AdaGradHyperparameter # NOQA
 _default_hyperparam.lr = 0.001
-_default_hyperparam.eps = 1e-8
+_default_hyperparam.batch_size_factor = 1
```
Comment (niboshi) [inline chainer/optimizers/ada_grad.py]: I think the default `batch_size_factor` can be set in `optimizer.Hyperparameter()` base class.
`eps` should not be removed.

File: chainer/optimizers/ada_grad.py [hunk 3]
```diff
@@ -39,13 +39,16 @@ class AdaGradRule(optimizer.UpdateRule):
     """
     _kernel = None
 
-    def __init__(self, parent_hyperparam=None, lr=None, eps=None):
+    def __init__(self, parent_hyperparam=None, lr=None, eps=None,
+                 batch_size_factor=None):
```
Comment (niboshi) [inline chainer/optimizers/ada_grad.py]: `batch_size_factor` can be omitted? Is there any use case to set it in the constructor?

File: chainer/optimizers/ada_grad.py [hunk 4]
```diff
@@ -90,13 +94,16 @@ class AdaGrad(optimizer.GradientMethod):
 
     """
 
-    def __init__(self, lr=_default_hyperparam.lr, eps=_default_hyperparam.eps):
+    def __init__(self, lr=_default_hyperparam.lr, eps=_default_hyperparam.eps,
+                 batch_size_factor=_default_hyperparam.batch_size_factor):
```
Comment (niboshi) [inline chainer/optimizers/ada_grad.py]: Ditto. No need to set it in the constructor?

File: chainer/optimizers/ada_grad.py [hunk 5]
```diff
@@ -90,13 +94,16 @@ class AdaGrad(optimizer.GradientMethod):
...
         super(AdaGrad, self).__init__()
         self.hyperparam.lr = lr
         self.hyperparam.eps = eps
+        self.hyperparam.batch_size_factor = batch_size_factor
 
     lr = optimizer.HyperparameterProxy('lr')
     eps = optimizer.HyperparameterProxy('eps')
+    batch_size_factor = optimizer.Hyperparameter('batch_size_factor')
```
Comment (niboshi) [inline chainer/optimizers/ada_grad.py]: This can be defined in the base `Optimizer` class.

File: chainer/optimizer.py (hunk 1 above)
Comment (niboshi) [reply chainer/optimizer.py]: I left some comments about the implementation.

File: chainer/optimizers/ada_grad.py (hunk 3 above)
Comment (emcastillo) [reply chainer/optimizers/ada_grad.py]: The `test_hyperparams` method get the list of hyperparameters for the optimizer and tries to instantiate it passing all the hyperparameters as constructor arguments
```
        default = self.optimizer.hyperparam.get_dict()
        for name, default_value in six.iteritems(default):
            self.create()
            self.assertEqual(self.get_hyperparam(name), default_value)
            new_value = default_value + 0.1
            self.create(**{name: new_value})
            self.assertEqual(self.get_hyperparam(name), new_value)
```
Solutions for this are to explicitly disable the batch_size_factor in the test.
or support batch_size_factor hyperparameter in the optimizer constructor.

I am up for the first one.

Comment (emcastillo) [reply chainer/optimizers/ada_grad.py]: Just added the complete implementation of this proposal following the suggestions I got yesterday. Thanks!

File: chainer/optimizer.py (hunk 1 above)
Comment (niboshi) [reply chainer/optimizer.py]: I'm just beginning to feel `batch_size_factor` might be too specific too.
How about `grad_scale_factor`, `loss_scale_factor` (perhaps confusing?), etc.?

Comment (emcastillo) [reply chainer/optimizer.py]: I think that `grad_scale_factor` or `loss_scale_factor` gives the wrong idea  that the parameter is used to alter the grads or the loss value somehow. I think that `updater_scale_factor` may be generic enough.

Comment (niboshi) [reply chainer/optimizer.py]: Updaters are just an instance of users of optimizers. I think it would be a reverse dependency to name it based on updaters.

Comment (emcastillo) [reply chainer/optimizer.py]: We have discussed offline and found that there are no better fitting alternatives than `grad_scale_{factor, compensator}` for this use case.

Comment (emcastillo) [reply chainer/optimizer.py]: hyperparameter renamed

**Comment by emcastillo** [comment]
Fixed minor errors

**Comment by niboshi** [comment]
As @beam2d told it might not worth risking breaking backward compatibility for a feature that is planned to be deprecated (ChainerMN is going to replace `MultiprocessParallelUpdater`).
How do you think?

**Comment by emcastillo** [comment]
How long is going to take until the replacement?
If its short term (1~3 months) I completely agree.
